src/storage/btree_load.c

```diff
@@ -4435,109 +4435,168 @@ xbtree_load_online_index (THREAD_ENTRY * thread_p, BTID * btid, const char *bt_n
 
-  /* Assign the snapshot to the scan_cache. */
-  scan_cache.mvcc_snapshot = builder_snapshot;
+      /* Start the online index builder. */
+      ret =
+	online_index_builder (thread_p, &btid_int, &hfids[cur_class], &class_oids[cur_class], n_classes, attr_ids,
+			      n_attrs, func_index_info, filter_pred, attrs_prefix_length, &attr_info, &scan_cache,
+			      unique_pk);
+      if (ret != NO_ERROR)
+	{
+	  break;
+	}
 
-  /* Demote the lock. */
-  ret = lock_demote_class_lock (thread_p, class_oids, new_lock, &old_lock);
-  if (ret != NO_ERROR)
-    {
-      goto error;
-    }
+      if (attr_info_inited)
+	{
+	  heap_attrinfo_end (thread_p, &attr_info);
 
-  /* Start the online index builder. */
-  ret = online_index_builder (thread_p, &btid_int, hfids, class_oids, n_classes, attr_ids, n_attrs,
-			      func_index_info, filter_pred, attrs_prefix_length, &attr_info, &scan_cache, unique_pk);
+	  if (filter_pred)
+	    {
+	      heap_attrinfo_end (thread_p, filter_pred->cache_pred);
+	    }
+	  if (func_index_info.expr)
+	    {
+	      heap_attrinfo_end (thread_p, func_index_info.expr->cache_attrinfo);
+	    }
 
-  // We shold recover the lock regardless of return code from online_index_builder.
-  // Otherwise, we might be doomed to failure to abort the transaction.
-  // We are going to do best to avoid lock promotion errors such as timeout and deadlocked.
+	  attr_info_inited = false;
+	}
 
-  // never give up
-  old_wait_msec = xlogtb_reset_wait_msecs (thread_p, LK_INFINITE_WAIT);
+      if (scan_cache_inited)
+	{
+	  heap_scancache_end (thread_p, &scan_cache);
+	  scan_cache_inited = false;
+	}
+      if (filter_pred != NULL)
+	{
+	  /* to clear db values from dbvalue regu variable */
+	  qexec_clear_pred_context (thread_p, filter_pred, true);
 
-  /* Promote the lock to SCH_M_LOCK */
-  lock_ret = lock_object (thread_p, class_oids, oid_Root_class_oid, SCH_M_LOCK, LK_UNCOND_LOCK);
-  if (lock_ret != LK_GRANTED)
-    {
-      // FIXME: What can we do??
-      assert (lock_ret == LK_GRANTED);
-    }
+	  if (filter_pred->unpack_info != NULL)
+	    {
+	      stx_free_additional_buff (thread_p, filter_pred->unpack_info);
+	      stx_free_xasl_unpack_info (filter_pred->unpack_info);
+	      db_private_free_and_init (thread_p, filter_pred->unpack_info);
+	    }
+	}
+
+      if (func_index_info.expr != NULL)
+	{
+	  (void) qexec_clear_func_pred (thread_p, func_index_info.expr);
+	  func_index_info.expr = NULL;
+	}
 
-  // reset back
-  (void) xlogtb_reset_wait_msecs (thread_p, old_wait_msec);
+      if (func_unpack_info != NULL)
+	{
+	  stx_free_additional_buff (thread_p, func_unpack_info);
+	  stx_free_xasl_unpack_info (func_unpack_info);
+	  db_private_free_and_init (thread_p, func_unpack_info);
+	}
+
+      if (BTREE_IS_UNIQUE (unique_pk))
+	{
+	  /* Check if we have a unique constraint violation for unique indexes. */
+	  ret =
+	    btree_online_index_check_unique_constraint (thread_p, btid_int.sys_btid, bt_name, &class_oids[cur_class]);
```

Comment:
Is it safe to check it here? I need more observations.

Reply:
The check should be outside loop, after promoting the lock.